algorithms/src/snark/marlin/ahp/ahp.rs

```diff
@@ -176,38 +176,66 @@ impl<F: PrimeField, MM: MarlinMode> AHPForR1CS<F, MM> {
         let eta_b = first_round_msg.eta_b;
         let eta_c = first_round_msg.eta_c;
         let batch_combiners = &first_round_msg.batch_combiners;
-        let prover::ThirdMessage { sum_a, sum_b, sum_c } = prover_third_message;
+        let prover::FourthMessage { sum_a, sum_b, sum_c } = prover_fourth_message;
 
         #[rustfmt::skip]
         let t_at_beta =
             eta_a * state.non_zero_a_domain.size_as_field_element * sum_a +
             eta_b * state.non_zero_b_domain.size_as_field_element * sum_b +
             eta_c * state.non_zero_c_domain.size_as_field_element * sum_c;
-        let r_b = state.third_round_message.as_ref().unwrap().r_b;
-        let r_c = state.third_round_message.as_ref().unwrap().r_c;
+        let r_b = state.fourth_round_message.as_ref().unwrap().r_b;
+        let r_c = state.fourth_round_message.as_ref().unwrap().r_c;
 
-        let beta = state.second_round_message.unwrap().beta;
+        let delta = state.second_round_message.unwrap().delta;
+        let beta = state.third_round_message.unwrap().beta;
         let gamma = state.gamma.unwrap();
+        let zeta = state.zeta.unwrap();
 
         let mut linear_combinations = BTreeMap::new();
 
         // Lincheck sumcheck:
+        let z_a_s = (0..state.batch_size)
+            .map(|i| {
+                let z_a_i = witness_label("z_a", i);
```

Comment:
I suggest instead of constructing the linear combination for the rowcheck as `(z_a_beta * z_b_beta - z_c_beta) * s_m`, we should construct it as `(z_a * z_b_beta - z_c) * s_m_beta`. This saves  proof size: for a batch size of `n`,we only need to provide 1 evaluation of `s_m_beta` and `n` evaluations of `z_b_beta`, as opposed to `n` evaluations each of `z_a_beta`, `z_b_beta`, and `z_c_beta`.